assets/app/scripts/services/data.js

```diff
@@ -504,14 +504,48 @@ angular.module('openshiftConsole')
     }
   };
 
-  DataService.prototype._watchWebsocketRetries = function(type, context, retry) {
+  DataService.prototype._addWatchWebsocketEvent = function(type, context, eventType) {
     var key = this._uniqueKeyForTypeContext(type, context);
-    if (retry === undefined) {
-      return this._watchWebsocketRetriesMap[key];
+    var events = this._watchWebsocketEventsMap[key];
+    if (!events) {
+      events = this._watchWebsocketEventsMap[key] = [];
+    } else if (events.length === 10) {
+      // Only keep 10 events. Shift the array to make room for the new event.
+      events.shift();
+    }
+
+    // Add the event to the end of the array with the time in millis.
+    events.push({
+      type: eventType,
+      time: Date.now()
+    });
+  };
+
+  DataService.prototype._isTooManyWatchWebsocketRetries = function(type, context) {
+    var key = this._uniqueKeyForTypeContext(type, context);
+    var events = this._watchWebsocketEventsMap[key];
+
+    // If there are less than 5 events, always retry.
+    if (!events || events.length < 5) {
+      return false;
     }
-    else {
-      this._watchWebsocketRetriesMap[key] = retry;
+
+    // If we've had more than 10 events in 30 seconds, stop.
+    // The oldest event is at index 0.
+    if (events.length >= 10 && (Date.now() - events[0].time) < (1000 * 30)) {
+      Logger.log("Too many websocket open or close events for type/context in the last 30 seconds", type, context, events);
+      return true;
+    }
+
+    // Make sure the last 5 events were not close events, which means the
+    // connection is not succeeding. This check is necessary if connection
+    // timeouts take longer than 6 seconds.
+    for (var i = events.length - 5; i < events.length; i++) {
```

Comment:
can this be made a positive check... would like to return true in the middle of the method and false at the end... makes structure clearer if we're testing for different conditions